### PR TITLE
Add a no-interface option to no-inheritance and no-multiple-inheritance

### PR DESCRIPTION
**Pull request by nfeignon**
In response to [Augur bounties](https://augur.net/bounties/): "Prohibit use of multiple inheritance but allow interfaces" and "Prohibit use of inheritance but allow interfaces".

Let me know if there's anything to fix.

### CODE REVIEW

**Comment by duaraghav8** [comment]
@nfeignon I had a discussion about these 2 rules with tinybike. They need to be merged with `no-inheritance` & `no-multiple-inheritance` respectively. How this works is:

- We will have `no-inheritance` that doesn't allow inheritance but does allow interfaces (default setting).
- This rule can be supplied an option to disallow interfaces (inside `.soliumrc.json`)
```
{
  "rules": {
    "security/no-inheritance": ["error", { "no-interface": true }]
    "security/no-multiple-inheritance": ["error", { "no-interface": true }]
  }
}
```
The same applies to `no-multiple-inheritance` (as seen in above eg)

Can you please augment the changes into the existing rules? Sorry to cause you the trouble, but this approach is much more elegant.

**Comment by nfeignon** [comment]
Updated my PR. Let me know if there's anything to fix!

File: rules/no-inheritance.js [hunk 1]
```diff
@@ -15,26 +15,69 @@ module.exports = {
             description: 'Discourage use of inheritance'
         },
 
-        schema: []
+        schema: [{
+            type: 'object',
+            properties: {
+                'no-interface': { type: 'boolean' }
+            }
```
Comment (duaraghav8) [inline rules/no-inheritance.js]: add `additionalProperties: false` after `properties` object. We want to keep schema as strict and narrowed as we can.
```
schema: [{
  type: 'object',
  properties: {...},
  additionalProperties: false
}]
```

File: test/no-inheritance.js [hunk 2]
```diff
@@ -13,28 +13,71 @@ var userConfig = {
     }
 };
 
+var userConfigNoInterface = {
```
Comment (duaraghav8) [inline test/no-inheritance.js]: please add 1 special test section and pass invalid options to the rule. This ensures that the schema defined for options is strictly followed and nothing unexpected cripples in.
add below tests
`"security/no-inheritance": ["error", { "no-interffff": true }]`
`"security/no-inheritance": ["error", {}]`
`"security/no-inheritance": ["error", { 1: 2 }]`
`"security/no-inheritance": ["error", { "no-interface": null }]`
`"security/no-inheritance": ["error", [true]]`
`"security/no-inheritance": ["error", { "no-interface": [true] }]`
`"security/no-inheritance": ["error", { "": false }]`
`"security/no-inheritance": ["error", { "no-interface": true, extraRandomProperty: "blah" }]`

The only configurations that should pass are
`"security/no-inheritance": ["error", { "no-interface": true }]`
`"security/no-inheritance": ["error", { "no-interface": false }]`
`"security/no-inheritance": ["error"]`

(you can add any other tests that you feel might be relevant)

**Comment by duaraghav8** [comment]
@nfeignon I've requested 2 changes in `no-inheritance`. They also apply to `no-multiple-inheritance`. Everything else is good.

File: test/no-inheritance.js (hunk 2 above)
Comment (nfeignon) [reply test/no-inheritance.js]: > "security/no-inheritance": ["error", {}]

is considered valid. I did not add it to the tests.

**Comment by nfeignon** [comment]
Updated the PR. Tell me if there's anything else.

**Pull request merged**